Added support for running GPU runtime tests in parallel.

mhong opened the pull request:
This is done by setting the TF config proto field
`ConfigProto.gpu_options.allow_growth` to true, via experimental C APIs.

Also removed obsolete logic in the compiler runtime, which was needed for eager
runtime support. This in turn simplifies the test setup.

Commit:
Added support for running GPU runtime tests in parallel.

This is done by setting the TF config proto field
ConfigProto.gpu_options.allow_growth to true, via experimental C APIs.

Also removed obsolete logic in the compiler runtime, which was needed for eager
runtime support. This in turn simplifies the test setup.

mhong commented:
This PR requires a change to TF for adding the experimental C APIs first. I'll wait for that to land in OSS TF, and then update the PR to use the new TF version.

mhong commented:
@marcrasi FYI -- this PR removed -DCPU in our CMakes runtime tests config.

dan-zheng commented on stdlib/private/TensorFlowUnittest/TensorFlowUnittest.swift:
```diff
@@ -53,51 +53,34 @@ public func printT(_ message: String) {
-  // is available.
-  // 2. TPU is mutually exclusive with {CPU, CUDA} (requiring different
-  // hardware).
+  // If the macro TPU is not specified, run the tests with CPU or GPU. Otherwise
```
It seems previously, CPU and GPU tests were both run with `-DCUDA`.
Is there a reason why now only GPU tests are run?

mhong replied:
This PR removes the runtime config distinction between CPU and GPU tests. At compile time, user uses -Xllvm -tf-target-gpu to specify whether they target CPU or GPU. The runtime is then agnostic of the CPU vs GPU device.

Unfortunately, it still needs to be aware of XLA/TPU for now. Revisiting that is outside the scope of this PR (I left a TODO there though).

dan-zheng commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -54,17 +54,16 @@ public func enableTPU(serverAddress: String? = nil, infeed: Bool = true) {
-  case gpu
-  /// TPU backend.
+  /// CPU or GPU execution.
+  case auto
```
If `cpu` and `gpu` execution modes are merged, is it still possible to run CPU tests if Swift is built with `-DCUDA`?

mhong replied:
Yes! With the new device placement design, if user does not specify `-Xllvm -tf-target-gpu`, the graph runs on CPU. In other words, we are not relying on TF-level soft placement anymore. Our generated graph ops always have explicit device placement.

Commit:
Addressed feedback from @mrry on the API design of TF_CreateConfig().

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -623,8 +637,9 @@ public final class _TensorComputation {
-    let context = _ExecutionContext.global
+    // Initialize global execution context if that's not yet done. It caches all
+    // our tensor programs.
+    let _ = _ExecutionContext.global
```
You can remove ‘let’ leaving just an underscore.

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -242,6 +237,23 @@ public final class _ExecutionContext {
+      _RuntimeConfig.executionMode == .xla ? 1 : 0,
+      _RuntimeConfig.gpuMemoryAllowGrowth ? 1 : 0)
+    TFE_ContextOptionsSetConfig(opts,
+                                self.tensorFlowConfig.pointee.data,
```
No need for self here and for other places.

mhong replied:
Done

mhong commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -355,11 +368,12 @@ fileprivate extension _ExecutionContext {
-        TF_EnableXLACompilation(sessionOptions, 1)
-      }
+      TF_SetConfig(sessionOptions,
+                   self.tensorFlowConfig.pointee.data,
```
Done

Commit:
Updated TF repo commit hash to pull in new changes on:
1. The experimental C APIs to enable GPU memory allocator growth, thus enabling
parallel gpu unit testing.
2. Allow a TF If op to produce 0 outputs, thereby fixing SR-8106.

Also addressed more comments from @rxwei.

mhong commented:
Test passed: https://ci-external.swift.org/view/Pull%20Request/job/swift-PR-TensorFlow-Linux/288/

dan-zheng commented:
Now that this PR is merged, I'll push build script support for GPU soon.

mhong commented:
Thanks Dan!
